GVFS/GVFS.Common/ProductUpgrader.cs

```diff
@@ -0,0 +1,489 @@
+            errorMessage = null;
+
+            try
+            {
+                Stream result = client.GetStreamAsync(GitHubReleaseURL).GetAwaiter().GetResult();
+
+                DataContractJsonSerializer serializer = new DataContractJsonSerializer(typeof(List<Release>));
+                releases = serializer.ReadObject(result) as List<Release>;
+                return true;
+            }
+            catch (HttpRequestException exception)
+            {
+                errorMessage = string.Format("Network error: could not connect to GitHub({0}). {1}", GitHubReleaseURL, exception.Message);
+                this.TraceException(exception, nameof(this.TryFetchReleases), $"Error fetching release info.");
+            }
+            catch (SerializationException exception)
+            {
+                errorMessage = string.Format("Parse error: could not parse releases info from GitHub({0}). {1}", GitHubReleaseURL, exception.Message);
+                this.TraceException(exception, nameof(this.TryFetchReleases), $"Error parsing release info.");
+            }
+
+            return false;
+        }
+
+        protected virtual void RunInstaller(string path, string args, out int exitCode, out string error)
+        {
+            ProcessResult processResult = ProcessHelper.Run(path, args);
+
+            exitCode = processResult.ExitCode;
+            error = processResult.Errors;
+        }
+
+        private static bool TryCreateDirectory(string path, out Exception exception)
+        {
+            try
+            {
+                Directory.CreateDirectory(path);
+            }
+            catch (IOException e)
+            {
+                exception = e;
+                return false;
+            }
+            catch (UnauthorizedAccessException e)
+            {
+                exception = e;
+                return false;
+            }
+
+            exception = null;
+            return true;
+        }
+
+        private bool TryRunInstallerForAsset(string name, out int installerExitCode, out string error)
+        {
+            error = null;
+            installerExitCode = 0;
+
+            bool installerIsRun = false;
+            string path;
+            string installerArgs;
+            if (this.TryGetLocalInstallerPath(name, out path, out installerArgs))
+            {
+                string logFilePath = GVFSEnlistment.GetNewLogFileName(GetLogDirectoryPath(), Path.GetFileNameWithoutExtension(path));
+                string args = installerArgs + " /Log=" + logFilePath;
+                this.RunInstaller(path, args, out installerExitCode, out error);
+
+                if (installerExitCode != 0 && string.IsNullOrEmpty(error))
+                {
+                    error = name + " installer failed. Error log: " + logFilePath;
+                }
+
+                installerIsRun = true;
+            }
+            else
+            {
+                error = "Could not find downloaded installer for " + name;

```

Comment:
>>It should have known that a file was missing after the download step. It should not have upgraded Git when it was guaranteed that upgrading GVFS would fail.

Agree. Added early exit, when either Git or GVFS asset installers are missing.

>>It failed to upgrade GVFS, and yet it still remounted the repos. We've now put the user into a state where the Git and GVFS versions do not match.

Updated to not auto re-mount when GVFS installation failed.

>> There was a long delay between "Mounting repositories...Succeeded" and "Error: Could not find". At least 10 seconds, if not longer, with nothing moving on the screen. Any idea what happened?

I couldn't repro this, but there is a step to delete the downloaded assets after Mounting Finished. Am wondering if something (like AV - am guessing) slowed the file deletion on your PC.